clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop3/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,418 @@
+        "Path %s doesn't start with scheme \"%s\".");
+
+    if (StringUtils.isBlank(virtualUri.toString())) {
+      throw new InvalidPathException("Uri which need be extracted cannot be null or empty.");
+    }
+
+    // remove first '/' symbol with empty string
+    String[] reservedDirs =
+        Arrays.stream(virtualUri.getPath().replaceFirst("/", "").split("/")).toArray(String[]::new);
+    Preconditions.checkArgument(
+        reservedDirs.length >= 3, "URI %s doesn't contains valid identifier", virtualUri);
+
+    return NameIdentifier.ofFileset(
+        metalakeName, reservedDirs[0], reservedDirs[1], reservedDirs[2]);
+  }
+
+  private FilesetContext getFilesetContext(Path virtualPath) {
+    NameIdentifier identifier = extractIdentifier(virtualPath.toUri());
+    Pair<Fileset, FileSystem> pair = filesetCache.get(identifier, this::constructNewFilesetPair);
+    Preconditions.checkState(
+        pair != null,
+        "Cannot get the pair of fileset instance and actual file system for %s",
+        identifier);
+    Path actualPath = getActualPathByIdentifier(identifier, pair, virtualPath);
+    return FilesetContext.builder()
+        .withIdentifier(identifier)
+        .withFileset(pair.getLeft())
+        .withFileSystem(pair.getRight())
+        .withActualPath(actualPath)
+        .build();
+  }
+
+  private Pair<Fileset, FileSystem> constructNewFilesetPair(NameIdentifier identifier) {
+    // Always create a new file system instance for the fileset.
+    // Therefore, users cannot bypass gvfs and use `FileSystem.get()` to directly obtain the
+    // Filesystem
+    try {
+      Fileset fileset = loadFileset(identifier);
+      URI storageUri = URI.create(fileset.storageLocation());
+      FileSystem actualFileSystem = FileSystem.newInstance(storageUri, getConf());
+      Preconditions.checkState(actualFileSystem != null, "Cannot get the actual file system");
+      return Pair.of(fileset, actualFileSystem);
+    } catch (IOException e) {
+      throw new RuntimeException(
+          String.format(
+              "Cannot create file system for fileset: %s, exception: %s",
+              identifier, e.getMessage()),
+          e);
+    } catch (RuntimeException e) {
+      throw new RuntimeException(
+          String.format(
+              "Cannot load fileset: %s from the server. exception: %s",
+              identifier, e.getMessage()));
+    }
+  }
+
+  private Fileset loadFileset(NameIdentifier identifier) {
+    Catalog catalog =
+        client.loadCatalog(NameIdentifier.ofCatalog(metalakeName, identifier.namespace().level(1)));
+    return catalog.asFilesetCatalog().loadFileset(identifier);
+  }
+
+  @Override
+  public URI getUri() {
+    return this.uri;
+  }
+
+  @Override
+  public Path getWorkingDirectory() {
+    return this.workingDirectory;
+  }
+
+  @Override
+  public void setWorkingDirectory(Path newDir) {
+    FilesetContext context = getFilesetContext(newDir);
```

Comment:
Do we should handle the concurrency cases for this method?

Reply:
Generally, the working dir does not change frequently, but referring to the implementation in `WebHdfsFileSystem`, we can add the `synchronized` keyword to this method